include/revng/Model/Importer/Binary/Options.h

```diff
@@ -0,0 +1,43 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <cstdint>
+#include <string>
+
+#include "llvm/ADT/ArrayRef.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Support/CommandLine.h"
+
+// This interface description makes it hard to allow the user to specify
+// the value for the search depth with the desired granularity.
+//
+// For now, I'll set it to reasonable default value of `3`, which means that
+// the debug information is searched in the binary itself (level `1`), its
+// direct dependencies (level `2`) and the dependencies of those libraries.
+//
+// TODO: tweak the `Libraries` value / extend the enum if the need for more
+//       granular control ever arises.
+enum class DebugInfoLevel { No = 0, Yes = 1, Libraries = 3 };
```

Comment:
1. Here you state the default is `3`, but it's not the case in the corresponding `cl::opt`.
2. Very strange to skip 2. I think the whole the "level" thing should be simplified to: 0) no, 1) only binary, 2) also libraries. And "also libraries" should be the default. I'm not sure what the distinction between the current 1 and 2 is, but I don't think we really care to expose that to the user.

cc/@dwarfyresearch

Reply:
Sorry I misled you, I'll rewrite the comment. We have three modes: (0) ignore debug info, (1) only load debug info from the binary and (>1) the depth of recursive dependencies we are content with looking at.

The comment here talks exclusively about choosing the value for the last (named as `Libraries`) option user is given. I feel like they don't really care about the specific depth we look at, so we can just choose a "sensible default", maybe 3 or 4 and leave it at that.

Reply:
> what the distinction between the current 1 and 2

It's the distinction between `2` and `3` actually. And the whole point of hard-setting `Libraries` value to one of these to spare the user (as in, the person who types in `revng pipeline --help` and then passes the value they liked in) this detail just providing them with a "sensible default" lookup depth.

Reply:
> And "also libraries" should be the default.

No strong opinion on this. @dwarfyresearch?

Reply:
This has been confusing for a while now. Please let's just have three modes: yes, no, libraries too.
Then, internally, the importer can do whatever it wants, but CLI options and `ImporterOptions` should just have the three simple options above, no holes.

Reply:
So I guess what you want is to just hide this comment closer to the implementation and that's it?

Reply:
>Very strange to skip 2. I think the whole the "level" thing should be simplified to: 0) no, 1) only binary, 2) also libraries. And "also libraries" should be the default. I'm not sure what the distinction between the current 1 and 2 is, but I don't think we really care to expose that to the user.

I think I agree with this. Let's use ```2``` for the ```Libraries``` and let it be the default. By doing that, the default will be searching for the debug info in the libraries on the first dependency level (so it means in the libraries the binary depends on directly).

Reply:
> I think I agree with this. Let's use 2 for the Libraries and let it be the default. By doing that, the default will be searching for the debug info in the libraries on the first dependency level (so it means in the libraries the binary depends on directly).

I feel like @aleclearmind meant it more aking to "remove explicit numbers from the enum and then in cpp file just do `Options.DebugInfo == DebugInfoLevel::Libraries ? WhateverNumberYouWant : 1`, as in, to hide it as an implementation detail.

Also, is there any real downsides of cranking they value really high up? Like `5` or `6`? Aside from the importer being somewhat slower?

Reply:
> > I think I agree with this. Let's use 2 for the Libraries and let it be the default. By doing that, the default will be searching for the debug info in the libraries on the first dependency level (so it means in the libraries the binary depends on directly).
> 
> I feel like @aleclearmind meant it more aking to "remove explicit numbers from the enum and then in cpp file just do `Options.DebugInfo == DebugInfoLevel::Libraries ? WhateverNumberYouWant : 1`", as in, to hide it as an implementation detail.
> 

I don't think we should explicitly enumerate things in the enum ```0, 1, 2```, no need for it for sure, but I was just emphasizing that it is strange to skip ```2``` in the enum. 

> Also, is there any real downsides of cranking they value really high up? Like `5` or `6`? Aside from the importer being somewhat slower?

Hmm, apart from it, I'd say no. Anyway, I am not sure how much sense would be there to invoke it with such a high level, but it is possible in theory that you can find a type there.

Reply:
> So I guess what you want is to just hide this comment closer to the implementation and that's it?

```c++
enum class DebugInfoLevel { No = 0, Yes = 1, Libraries = 2 };
```

Where `Yes` means what currently is `2` and `Libraries` means what currently is `3`.

And:

```c++
cl::opt<DebugInfoLevel> DebugInfo("debug-info",
                                  cl::desc(DescLevel),
                                  cl::value_desc("level"),
                                  cl::values(No, Yes, Lib),
                                  cl::cat(MainCategory),
                                  cl::init(DebugInfoLevel::Libraries));
```

Inspecting libraries deeper than those directly linked makes sense in very rare circumstances, we can expose such feature to the user in the future if needed.

Reply:
> `DebugInfoLevel`

Isn't the whole point of `Yes` to let somebody only load a single binary, disregarding any dependencies and stuff? I'd say that option is more valuable than differentiating between "only look at direct libraries" and "look at libraries a couple of levels in".

I'm just afraid that the feature might get lost for no reason. Also, if we're making `Libraries` the default, I think it's better to go the full mile and just call that `Yes`.

What do you think about something like:
```cpp
enum class DebugInfoLevel {
  No, // Ignore debug information
  Yes, // Load the debug information from the binary dynamic libraries it directly links
  BinaryOnly, // Ignore debug information in the dynamically linked libraries
  DeepSearch // Also try to look for the debug symbols in indirectly linked libraries, up to 5 levels of depth. This makes importing slower.
};
```

> DebugInfo

This is an [existing flag](https://github.com/revng/revng/blob/7361da36291db4b6fb822b8072145ac47d1e08df/include/revng/Model/Importer/Binary/BinaryImporterOptions.h#L11-L14) that I just moved under the same umbrella because it's close enough - and this way there's one less `inline` variable in a header

Reply:
As I mentioned, `DeepSearch` is not a feature we want:

> Inspecting libraries deeper than those directly linked makes sense in very rare circumstances, we can expose such feature to the user in the future if needed.

I'm fine with `No, IgnoreLibraries, Yes`, if you prefer.

Reply:
Done, please take a look.

Reply:
Looks great!